src/applications/Applications.DicomLoader/Loader.cs

```diff
@@ -0,0 +1,165 @@
+﻿#nullable enable
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using DicomTypeTranslation;
+using FellowOakDicom;
+using MongoDB.Bson;
+using MongoDB.Driver;
+using SharpCompress.Readers;
+using Smi.Common.Messages;
+using Smi.Common.MongoDB;
+
+namespace Applications.DicomLoader;
+
+public static class Loader
+{
+    private static readonly object _flushLock=new ();
+    private static int _fileCount;
+    private static readonly ConcurrentDictionary<string,SeriesMessage> _seriesList=new ();
+    private static readonly Stopwatch _timer = Stopwatch.StartNew();
+
+    private static SeriesMessage LoadSm(string id, string directoryName, DicomDataset ds, string studyId)
+    {
+        // Try loading from Mongo in case we were interrupted previously
+        var b=SeriesStore.Find(new BsonDocument("SeriesInstanceUID", id)).FirstOrDefault();
+        return b ?? new SeriesMessage
+        {
+            DirectoryPath = directoryName,
+            DicomDataset = DicomTypeTranslater.SerializeDatasetToJson(ds),
+            ImagesInSeries = 1,
+            SeriesInstanceUID = id,
+            StudyInstanceUID = studyId
+        };
+    }
+    private static SeriesMessage IncSm(string _,SeriesMessage sm)
+    {
+        sm.ImagesInSeries++;
+        return sm;
+    }
+    /// <summary>
+    /// Write the pending Series data out to Mongo
+    /// </summary>
+    public static void Flush(bool force=false)
+    {
+        lock (_flushLock)
+        {
+            if (!force && _seriesList.Count < 100)
+                return;
+            if (!_seriesList.IsEmpty)
+                SeriesStore?.InsertMany(_seriesList.Values);
+            _seriesList.Clear();
+        }
+    }
+
+    public static void Report()
+    {
+        if (_timer.ElapsedMilliseconds == 0) return;
+        Console.WriteLine($"Processed {_fileCount} files in {_timer.ElapsedMilliseconds}ms ({1000*_fileCount/_timer.ElapsedMilliseconds} per second)");
+    }
+
+    private static readonly byte[] _dicomMagic = Encoding.ASCII.GetBytes("DICM");
+
+    private static void Process(FileInfo fi, IMongoCollection<BsonDocument> iStore, CancellationToken ct)
+    {
+        var bBuffer=new byte[132];
+        var buffer = new Span<byte>(bBuffer);
+        using var fileStream = File.OpenRead(fi.FullName);
+        if (fileStream.Read(buffer) == 132 && buffer[128..].SequenceEqual(_dicomMagic))
```

Comment:
Maybe we should let `.dcm` files in anyway? is there any chance that any dicom file misses out this header?

Reply:
Not if it wants to get into fo-dicom without an exception - fo-dicom throws "unknown file format" if the DICOM 3 signature is missing.